### PR TITLE
Custom bundling

### PR DESCRIPTION
**Pull request by P0lip**
https://github.com/stoplightio/platform-internal/issues/4468

TODO:
- [x] docs :laughing: 
- [x] better typings + jsdoc

Can be tried in the repo I sent on Slack.

### CODE REVIEW

File: lib/bundle/index.js [hunk 1]
```diff
@@ -131,12 +168,13 @@ function inventory$Ref ($refParent, $refKey, path, pathFromRoot, indirections, i
     circular: pointer.circular,   // Is this $ref pointer DIRECTLY circular? (i.e. it references itself)
     extended,           // Does this $ref extend its resolved value? (i.e. it has extra properties, in addition to "$ref")
     external,           // Does this $ref pointer point to a file other than the main JSON Schema file?
-    indirections,   // The number of indirect references that were traversed to resolve the value
+    indirections,   // The number of indirect references that were traversed to resolve the value,
```
Comment (wmhilton) [inline lib/bundle/index.js:171]:
```suggestion
    indirections,   // The number of indirect references that were traversed to resolve the value
```

File: lib/bundle/index.js [hunk 2]
```diff
@@ -211,16 +256,33 @@ function remap (inventory) {
...
+        let parsedHash = safePointerToPath(entry.hash);
+        let value = get(subschema, `#/${parsedHash.length === 1 ? parsedHash[0] : parsedHash.slice(0, parsedHash.length - 1).join("/")}`);
+        set(schema, customRoots[entry.file][entry.hash], parsedHash.length === 1 ? value : value[parsedHash[parsedHash.length - 1]]);
+        delete value[parsedHash[parsedHash.length - 1]];
+      }
+
+      pathFromRoot = entry.pathFromRoot;
+      hash = customRoots[entry.file][entry.hash];
+      entry.$ref.$ref = entry.pathFromRoot;
+    }
```
Comment (wmhilton) [inline lib/bundle/index.js:275]: Can you add a comment (or more) here? I can't figure out what it's doing. Is this where it's rewriting the pointers to point at the bundled schemas?

File: lib/bundle/util/object.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    curObj = curObj[segment];
+
+    if (typeof curObj !== "object" || curObj === null) {
+      curObj = prevObj[segment] = {};
+    }
+  }
+
+  if (!(tokens[tokens.length - 1] in curObj)) {
+    curObj[tokens[tokens.length - 1]] = value;
+  }
```
Comment (wmhilton) [inline lib/bundle/util/object.js:56]: Just checking - is silently failing to set the value if it already exists intended?

File: lib/bundle/util/suggestName.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+  }));
+}
+
+function suggestName (root, takenKeys, name) {
+  let suggestedName = name;
+  let i = 2;
+  while ((root && suggestedName in root) || takenKeys.has(suggestedName)) {
+    suggestedName = `${name}_${i++}`;
+    if (i > MAX_ATTEMPTS) {
+      throw new Error("Too many attempts");
```
Comment (wmhilton) [inline lib/bundle/util/suggestName.js:25]:
```suggestion
      throw new Error(`suggestName: MAX_ATTEMPTS exceeded. Names ${name}_2 through ${name}_${MAX_ATTEMPTS} already taken.`);
```

**Review by wmhilton — APPROVED** [review]
Looks good. Quite a handful of code, but not more than there needs to be.

File: lib/bundle/util/object.js (hunk 3 above)
Comment (P0lip) [reply lib/bundle/util/object.js:56]: Yeah. We don't want to override it.

File: lib/bundle/index.js (hunk 2 above)
Comment (P0lip) [reply lib/bundle/index.js:275]: Good point. I'll leave a comment explaining what's happening here.

Comment (P0lip) [reply lib/bundle/index.js]: @wmhilton added! LMK what you think.

**Comment by P0lip** [comment]
@marbemac ping, I want to have this in assuming the output meets the expectations.

**Review by marbemac — APPROVED** [review]
🔥 let's get this into v1 and v2 - hopefully just works all around 🤞 . Likely to be some small things to tweak/fix that come out of it, but once that's done we can close this saga out! Nice work.

**Pull request merged**